docs/src/main/asciidoc/se/injection.adoc

```diff
@@ -0,0 +1,677 @@
+
+To declare an asynchronous observer use annotation `Event.AsyncObserver` instead of `Event.Observer`.
+
+[source,java]
+.Asynchronous Event Observer Example
+----
+include::{sourcedir}/se/inject/EventsExample.java[tag=snippet_5, indent=0]
+----
+
+== Programmatic Lookup
+
+There are two primary ways to access services from the service registry:
+
+- *Dependency Injection* – Services are injected automatically at defined injection points.
+- *Programmatic Lookup* – Services are retrieved manually by accessing the service registry directly.
+
+If you want to use programmatic lookup, there are two main ways to access a `ServiceRegistry` instance:
+
+- Create a new ServiceRegistry instance and search for the desired service.
+- Inject the existing ServiceRegistry instance currently used by Helidon and retrieve services from it.
+
+To create a registry instance:
+
+[source,java]
+----
+include::{sourcedir}/se/inject/ServiceRegistryExample.java[tag=snippet_1, indent=0]
+----
+
+Note that all instances are created lazily, meaning the service registry does nothing by default.
+If a service performs any actions during construction or post-construction,
+you must first retrieve an instance from the registry to trigger its initialization.
+
+Special registry operations:
+
+- `List<ServiceInfo> lookupServices(Lookup lookup)` - get all service descriptors that match the lookup
+- `Optional<T> get(ServiceInfo)` - get an instance for the provided service descriptor
+
+The common registry operations are grouped by method name. Acceptable parameters are described below.
+
+Registry methods:
+
+- `T get(...)` - immediately get an instance of a contract from the registry; throws if implementation not available
+- `Optional<T> first(...)` - immediately get an instance of a contract from the registry; there may not be an implementation
+available
+- `List<T> all(...)` - immediately get all instances of a contract from the registry; result may be empty
+- `Supplier<T> supply(...)` - get a supplier of an instance; the service may be instantiated only when `get` is called
+- `Supplier<Optional<T>> supplyFirst(...)` - get a supplier of an optional instance
+- `Supplier<List<T>> supplyAll(...)` - get a supplier of all instances
+
+Lookup parameter options:
+
+- `Class<?>` - the contract we are looking for
+- `TypeName` - the same, but using Helidon abstraction of type names (may have type arguments)
+- `Lookup` - a full search criteria for a registry lookup
+
+== Startup
+
+Helidon provides a Maven plugin (`io.helidon.service:helidon-service-maven-plugin`, goal `create-application`) to generate
+build time bindings, that can be used to start the service registry without any classpath discovery and reflection.
```

Comment:
`ApplicationBinding` provides build time bindings of service providers to injection points - this is to avoid lookup at runtime.
It does NOT elimiminate classpath discovery and reflection. To avoid that you need to use the generated `ApplicationMain` class.

Also classpath discovery does not mean classpath scanning (it sounds a bit bad in here) - we lookup all `service.registry` and `service.loader` resources from the classpath - we are not scanning for them (i.e. using only Java features that do not depend on how the classpath is created, so it works in native image as well)

Not sure how much of this should be in the docs, but it seems  a bit misleading right now.